File: Algorithm/BOJ/2667.cpp
//https://www.acmicpc.net/problem/2667
//http://1ilsang.blog.me/221330651097

#include <cstdio>
#include <vector>
#include <algorithm>
using namespace std;

int n;
vector<int> v;
int area[26][26];
int check[26][26];
int dr[4] = { 0,1,0,-1 };
int dc[4] = { 1,0,-1,0 };

int dfs(int r, int c) {
	int ret = 1;
	check[r][c] = 1;
	for (int i = 0; i < 4; i++) {
		int nr = r + dr[i];
		int nc = c + dc[i];
		if(0<= nr && 0 <= nc && nr < n && nc < n && check[nr][nc] == 0 && area[nr][nc] == 1){
			ret += dfs(nr, nc);
		}
	}
	return ret;
}

int main(void) {
	scanf("%d", &n);
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			scanf("%1d", &area[i][j]);
		}
	}
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (area[i][j] == 1 && check[i][j] == 0) {
				int tmp = dfs(i, j);
				if (tmp) {
					v.push_back(tmp);
				}
			}
		}
	}
	sort(v.begin(), v.end());
	printf("%d\n", v.size());
	for (int i = 0; i < v.size(); i++) {
		printf("%d\n", v[i]);
	}
	return 0;
}

File: Algorithm/BOJ/java/2667.java
//https://www.acmicpc.net/problem/2667

import java.io.*;
import java.util.*;

public class Main {
    private static int n;
    private static int[][] area, chk;
    private static int[] dr = {0, 1, 0, -1};
    private static int[] dc = {1, 0, -1, 0};

    private static int dfs(int r, int c) {
        int ret = 1;
        chk[r][c] = 1;
        for (int i = 0; i < 4; i++) {
            int nr = r + dr[i];
            int nc = c + dc[i];
            if (nr >= 0 && nc >= 0 && nr < n && nc < n && area[nr][nc] == 1 && chk[nr][nc] == 0) {
                ret += dfs(nr, nc);
            }
        }
        return ret;
    }

    public static void main(String[] args) throws IOException {
        ArrayList<Integer> list = new ArrayList<>();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        area = new int[n][n];
        chk = new int[n][n];

        for (int i = 0; i < n; i++) {
            char[] tmp = br.readLine().toCharArray();
            for (int j = 0; j < n; j++) {
                area[i][j] = tmp[j] - '0';
            }
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (area[i][j] == 1 && chk[i][j] == 0) {
                    list.add(dfs(i, j));
                }
            }
        }

        Collections.sort(list);
        System.out.println(list.size());
        for (int i = 0; i < list.size(); i++) System.out.println(list.get(i));
    }
}
